### DISCUSSION TITLE
How to properly reference IKVM-created DLLs in depending projects properly?

### DISCUSSION THREAD

**Discussion by nico-csp** [ikvmnet/ikvm#418] [Q&A] [answered]
# Status Quo

* Project A references a JAR.
* During build, the DLL is created automatically; `IkvmReference` does the (good) job.

# Next Level

* Project B depends on project A.
* Some of the classes in project A's JAR-DLL need to be returned into project B.

# Issue

Visual Studio complains that references are missing. So what happends automatically to 100% in project A, obviously requires some hands-on in project B.

# Best Solution?

What would be the best solution in that case?

Of course, I could simply create a direct reference to the file in project B. But it does not feel right. Isn't there some kind of automatics provided by IKVM? So that referencing project A is just enough?

**Comment by wasabii** [comment]
Got to add the reference to each project, just like normal assembly references. They are not transitive.

MavenReferences are.

**Reply by nico-csp** [reply]
I am not sure, whether I completely understood what you meant exactly; sorry.

Do you mean I have to add `IkvmReference` to each project, where I need references to that JAR, which is located in project A?

**Reply by wasabii** [reply]
Yes.

**Reply by nico-csp** [reply]
Ok, thank You for this clarification.

**Reply by wasabii** [reply]
It operates the same as `<Reference>`. Standard assembly references. It is not referenced transitively. Transitive means inherited by projects.

PackageReference is transitive. And, in that spirit, so is MavenReference.

**Reply by nico-csp** [reply]
Thanks a lot for this clarification.

Sorry for reopening the discussion, but I guess my next questions are directly belonging to this topic:

In order to keep it as simple as possible, I tried to just put the generated DLLs into another project, but Visual Studio refuses to add them as reference ("invalid").

Also I tried to create a NuGet-Package, but the problem is the same as described above (of course). In the future, I would like to have the classes in the JAR being provided as NuGet. What would be the correct/ideal way to provide those classes in the JAR (concrete JAR files, no Maven reference) as part of a NuGet package? Is there some documentation or tutorial, which I might have overseen (in that case: sorry for bothering you!)?

**Reply by wasabii** [reply]
It is recommended you do not do that, but have each project generate their own copy from the JAR file. This is what MavenReference accomplishes.

**Reply by nico-csp** [reply]
So there is no way of distributing the JAR files as part of a NuGet-Package and let it work by referencing the NuGet? Sorry for being a pain ... here... :-) Just need to make sure that I got it right. I embedded the JAR files in the project and in the NuGet, so the consuming projects should have all resources (JAR, DLL, IVKM).

I tried to figure out, whether I can somehow let `MavenReference` point to a local package file, because my JAR files will not be uploaded to any Maven-Repository. I read https://github.com/ikvmnet/ikvm-maven#readme but could not see, how to set the reference to local files.

Is there a simple example anyhwere, how to create a NuGet with JARs and `MavenReference` (all included) so that consuming projects can directly access the class in the JARs?

Some more details about my situation:
* `j1.jar` depends on `j2.jar`.
* Both are included in the "providing" project.
* This project contais the version information for the JAR files plus the reference information (dependency) -- so all meta data.
* The JAR files are not uploaded to any repository server, but are only locally stored in the project.
* It would not be ideal to have all the meta data for the JAR files to be copied into each consuming project.
* Would be nice to have a local package with the JAR files plus, their meta data so it can be referenced by any other project as simple as a NuGet package.

If `MavenReference` was the solution, I would be glad about a hint how to accomplish that. Any entry-point to a documentation, example or whatever would do it... Thanks in advance for your patience and support!

**Reply by wasabii** [reply]
So here's the thing. We don't guarantee compatibility between any generated DLL and IKVM runtime. Between versions we may break that interface. In fact we are likely to do so for the foreseeable future.

We don't approve of uploading generated assembly to public nuget: it messee up dependencies. This is outlined in the README.

But if this is private nuget, with private JAR files, well then we don't care. But IkvmReference has no provisions for this. That said, you can probably get it working pretty automatically using IKVM.Net.SDK. I believe we added support for importing archives in that.

Also, DLLs aren't portable between Core and Framework.

**Reply by nico-csp** [reply]
I got your point and fully agree. Since the JAR files should only be used internally, we can accept the risk of broken interfaces in the future.

* We only would like to be able to use JAR files of our JAVA project directly in our .NET projects.
* However, it should be easy to set up in each .NET project; thus, it is out of question to modify each `.csproj`-file to mention all JAR files plus their dependencies.

It would be perfect to being able to create a package containing the JAR files, their dependencies and the reference to IKVM so that each .NET project is able to create the "JAR-DLLs" themselves using IKVM.

As far as I learned it so far (I might be terribly wrong), it is impossible as long as the JAR files cannot be uploaded to Maven repository.

The limitations of IKVM.Net.SDK as written in its readme make it impossible for us to use it either.

The solution, which seems to work in a first rough test was to directly include the generated DLLs in the `.csproj` file for having them included in the NuGet-package. Example for two JAR-DLLs:

```
	<ItemGroup>
	  <Content Include="bin\Release\DOTNETVERSION\jarlib1.dll">
	    <Pack>true</Pack>
	    <PackagePath>\lib\DOTNETVERSION</PackagePath>
	  </Content>
	  <Content Include="bin\Release\DOTNETVERSION\jarlib2.dll">
	    <Pack>true</Pack>
	    <PackagePath>\lib\DOTNETVERSION</PackagePath>
	  </Content>
	</ItemGroup>
```

replacing `DOTNETVERSION` by whatever is necessary.

Since IKVM-packages are referenced by version, I do not see a severe risk of broken interfaces.

The solution is not perfect (I don't like it much) -- but it works. We will use it unless pointed to alternatives, which might be better.

Thank you for your support, Jerome! :-)

**Reply by wasabii** [reply]
>The limitations of IKVM.Net.SDK as written in its readme make it impossible for us to use it either.

Such as?

>Since IKVM-packages are referenced by version, I do not see a severe risk of broken interfaces.

The problem ends up being version unification.  Say you have three .NET projects: A, B and C. A depends on B and C. Imagine B depends on a Third Party Library (T). Imagine T uses IKVM. T however gets updated, and currently has a dependency on IKVM version 8.5.

Let's say C is your project. It generates your JAR files. It depends on IKVM 8.4.

Which version of IKVM should A use?

Standard Project/PackageReference unification will result in A depending on 8.5. Even though you told it to depend on IKVM 8.4, and you generated your JAR files against 8.4: at build time, project A will be upversioned to 8.5 because A depends on B depends on T depends on 8.5. So, 8.5 will be what is run, and it will be run using your 8.4 assemblies, which won't be compatible.

MavenReference takes care of this exactly how Project and PackageReference takes care of it: transitively communicating resolved versions across project/package boundaries, and applying unification rules, so that the version A runs will be 8.5. It then compiles assemblies, on the fly, for 8.5. Even if those assemblies are distributed by third party packages. Basically: everything gets recompiled for the actual version of IKVM that would be in use at runtime.

MavenReference results in every downstream project having it's OWN IkvmRefernce: thus, each project compiles a version of the JAR that is compatible with the version of IKVM that was resolved by Nuget after unification. The same as if you put IkvmReference in each project: each project would compile it's own: against it's own version of IKVM.

So it's not impossible to do what you're doing. But there's a very good reason we don't do it out of the box.

**Reply by wasabii** [reply, marked as answer]
This is why generally my recommendation would be to publish your jar files to an internal Maven repo. Same as you do your internal nuget packages. In fact, most nuget repositories people use support Maven just fine. Azure DevOps, GitHub packages, etc, all support publishing Maven packages.

**Reply by nico-csp** [reply]
At least in my current situation, it would not be a problem. But let me try the `MavenReference` tag.

I created a local Maven-Repository for testing. How can I make `MavenReference` picking up the JAR files from there?

There is a `~/.m2/settings.xml` pointing to that local repository, but `MavenReference` does not seem to use it and looks in `maven.org`.

I have no experience with JAVA and Maven; so I am not sure, whether the issue is there or somewhere else. Could it be that I have to put something like a `MavenRepository`-Tag into my `csproj` file?

**Reply by nico-csp** [reply]
Ok, I searched in the code and found some hints to continue searching the internet and found #395. I will check that.

**Reply by nico-csp** [reply]
Ok, I created a local Maven-Repository-Directory, referenced it in the `.csproj`-File, put the JAR-files with appropriate BOM-files into that local directory structure and used `MavenReference`.

It works fine; my first tests did not succeed due to lack of knowledge about JAVA and Maven. Meanwhile I fixed all this issues here and there...

Thanks once more for your patience. Guess, we can really close this thread now. :-)
